### PR TITLE
config: adjust OnlineDDL in subtask config

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve? 
close https://github.com/pingcap/dm/issues/2034

### What is changed and how it works?
add AdjustOnlineDDL in upgrade

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: dm/config/subtask.go [hunk 1]
```diff
@@ -429,3 +429,11 @@ func (c *SubTaskConfig) Clone() (*SubTaskConfig, error) {
 
 	return clone, nil
 }
+
+// AdjustOnlineDDL adjust OnlineDDL through OnlineDDLScheme.
+func (c *SubTaskConfig) AdjustOnlineDDL() {
+	if c.OnlineDDLScheme == GHOST || c.OnlineDDLScheme == PT {
+		c.OnlineDDL = true
+	}
+	c.OnlineDDLScheme = ""
```
Comment (lance6716) [inline dm/config/subtask.go:438]: the original value should not be changed, to support downgrading

File: pkg/upgrade/upgrade.go [hunk 2]
```diff
@@ -275,3 +276,30 @@ func upgradeToVer3(ctx context.Context, cli *clientv3.Client) error {
 func upgradeToVer4(cli *clientv3.Client, uctx Context) error {
 	return nil
 }
+
+// upgradeToVer5 does upgrade operations from Ver4 to Ver5 (v2.0.7) to update online-ddl from online-ddl-scheme in configure.
```
Comment (lance6716) [inline pkg/upgrade/upgrade.go:280]:
```suggestion
// upgradeToVer5 does upgrade operations from Ver4 to Ver5 (v2.1.0) to update online-ddl from online-ddl-scheme in configure.
```

OK to not change, we may change the release version in future 😂

File: pkg/upgrade/upgrade_test.go [hunk 3]
```diff
@@ -162,3 +163,76 @@ func (t *testForEtcd) TestUpgradeToVer3(c *C) {
...
+			OnlineDDL:       true,
+			OnlineDDLScheme: "",
+		},
+		{
+			OnlineDDL:       false,
+			OnlineDDLScheme: "",
+		},
+	}
+	for i := range inputCfgs {
+		fmt.Println(i)
```
Comment (lance6716) [inline pkg/upgrade/upgrade_test.go:219]:
```suggestion
```

File: pkg/upgrade/upgrade.go [hunk 4]
```diff
@@ -275,3 +276,30 @@ func upgradeToVer3(ctx context.Context, cli *clientv3.Client) error {
...
+	}
+
+	for task, m := range uctx.SubTaskConfigs {
+		for sourceID, subCfg := range m {
+			subCfg.AdjustOnlineDDL()
+			value, err := subCfg.Toml()
+			if err != nil {
+				return err
+			}
+			key := common.UpstreamSubTaskKeyAdapter.Encode(sourceID, task)
```
Comment (lance6716) [inline pkg/upgrade/upgrade.go:297]: I suddenlt realized that the upgrading actions can't make sure that DM-worker will see the new config. If DM-worker finish upgrading too quickly, at the time of its starting up, DM-master has not run here so task in etcd is not changed.

Maybe we should change the upgrading logic in TiUP 🤔

File: dm/config/subtask.go (hunk 1 above)
Comment (glorv) [reply dm/config/subtask.go:438]: +1

**Comment by glorv** [comment]
I think it's better to do an `adjust` to the config struct after there are built so we can avoid doing the complex upgrading logic.

File: dm/config/subtask.go (hunk 1 above)
Comment (okJiang) [reply dm/config/subtask.go:438]: It does not affect the original configure...

File: pkg/upgrade/upgrade.go (hunk 4 above)
Comment (lichunzhu) [reply pkg/upgrade/upgrade.go]: Some possible solutions:
1. Let worker do adjust job again.
2. If a master finds that its own version info is greater than version Info in etcd, transfer leader to itself immediately and boostrap before this dm-master provides service.
I think we can do this in another PR. It's hard and inappropriate for TiUP to handle DM's own upgrade logic.

Comment (lichunzhu) [reply pkg/upgrade/upgrade.go]: I have checked the logic in dm-worker. dm-worker will adjust subtask config again when dm-worker reads configuration from etcd. So maybe we don't have such problem?

**Review by lichunzhu — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Pull request merged**